app/controllers/admin/ride_uploads_controller.rb

```diff
@@ -0,0 +1,139 @@
+class Admin::RideUploadsController < Admin::AdminApplicationController
+  before_action :set_ride_upload, only: [:show, :edit, :update, :schedule, :destroy]
+  before_action :set_ride_zone, only: [:index, :show, :new, :edit, :create, :update, :destroy]
+
+  # GET /ride_uploads
+  def index
+    @ride_uploads = RideUpload.where( ride_zone_id: @ride_zone.id )
+  end
+
+  # GET /admin/ride_zones/ca-25/ride_uploads/
+  def show
+    @failed_potential_rides = @ride_upload.failed_potential_rides
+  end
+
+  # GET /admin/ride_zones/ca-25/ride_uploads/new
+  def new
+    @ride_upload = RideUpload.new
+  end
+
+  # GET /admin/ride_zones/ca-25/ride_uploads/1/edit
+  def edit
+  end
+
+  # POST /admin/ride_zones/ca-25/ride_uploads
+  def create
+    @ride_upload = RideUpload.new(ride_upload_params)
+    @ride_upload.user = current_user
+    @ride_upload.ride_zone = @ride_zone
+    
+    # TODO: Should a radius check be added to PotentialRide, and should it be made geolocatable?
+    if @ride_upload.save
+      row_number = 0
+      total_rows = 0
+      successful_rows = 0
+      CSV.foreach( @ride_upload.csv_on_disk, headers: true ) do |row|
+        row_number += 1
+        potential_ride = PotentialRide.create(ride_zone: @ride_zone, ride_upload: @ride_upload, status: 0, row_number: row_number)
+        if potential_ride.populate_from_csv_row(row)
+          successful_rows += 1 
+        else
+          potential_ride.fail_because("Failed to populate csv row #{row_number}")
+        end
+        total_rows += 1
+        potential_ride.save
+      end
+      
+      @ride_upload.total_rows = total_rows
+      @ride_upload.successful_rows = successful_rows
+      @ride_upload.queued!
+      @ride_upload.save
+      
+      redirect_to admin_ride_zone_ride_upload_path(@ride_zone.slug, @ride_upload), notice: 'Ride upload was successfully created, you still need to schedule the rides.'
+    else
+      render :new
+    end
+  end
+  
+  # POST /admin/ride_zones/ca-25/ride_uploads/1
+  # Go through all the PotentialRides associated with this RideUpload
+  # If any can't be processed, fail the whole thing and don't process any
+  # If all look like they can be scheduled, do so
+  def schedule
+    fail_count = 0
+    @ride_upload.potential_rides.each do |potential_ride|
```

Comment:
Also a very long block to have in an action.

Reply:
Yes, agreed. Left in a note for refactoring, but I don't have clarity right now about how it should be done, so in the interest of expediency, if you cool with it, we'll leave as is.